Optimize insert performance of metrics and tag_id lookup

atanasovskib opened the pull request:
A refactor was done that optimized several things:
1. Metrics are organized by measurement and schema validation is done
   only once instead of every metric
2. Metric values are batched and inserted with binary COPY
3. Tag IDs are now cached with a LRU strategy

### Required for all PRs:

- [ ] Signed [CLA](https://influxdata.com/community/cla/).
- [ ] Associated README.md updated.
- [ ] Has appropriate unit tests.

Commit:
Optimize insert performance of metrics and tag_id lookup

A refactor was done that optimized several things:
1. Metrics are organized by measurement and schema validation is done
   only once instead of every metric
2. Metric values are batched and inserted with binary COPY
3. Tag IDs are now cached with a LRU strategy

RobAtticus commented on plugins/outputs/postgresql/utils/utils.go:
```diff
@@ -0,0 +1,158 @@
+func DerivePgDatatype(value interface{}) PgDataType {
+	switch value.(type) {
+	case bool:
+		return "boolean"
```
pretty much all of these things should be constants, like `pgBool`, `pgInt8`, etc

RobAtticus commented on plugins/outputs/postgresql/utils/utils.go:
```diff
@@ -0,0 +1,158 @@
+	case bool:
+		return "boolean"
+	case uint64:
+		return "int8"
```
can use `fallthrough` and just return “int8” in the last one?

atanasovskib replied:
compiler doesn't allow fallthrough when switching types

RobAtticus commented on plugins/outputs/postgresql/utils/utils.go:
```diff
@@ -0,0 +1,158 @@
+		return containThis == "serial"
+	}
+	if canThis == "float8" {
+		return containThis == "int4" || containThis == "int8"
```
can `float8` really contain `int8` in PG? `float8` can only represent a subset of integers in `int8` so I’m surprised they let you do that conversion

atanasovskib replied:
My bad, it can't be stored only up to 2^53

RobAtticus commented on plugins/outputs/postgresql/utils/utils.go:
```diff
@@ -0,0 +1,158 @@
+
+// QuoteLiteral returns a sanitized string safe to use in sql as a string literal
+func QuoteLiteral(name string) string {
+	return "'" + strings.Replace(name, "'", "''", -1) + "'"
```
are there not helper functions in the pq or sql libs to do this for you? seems a bit risky to write our own sql sanitizers

atanasovskib replied:
nothing found in the database/sql package but the pgx quoteIdentifier uses the same implementation for replacing " 
```
return `"` + strings.Replace(s, `"`, `""`, -1) + `"`
```

RobAtticus replied:
Fair enough

RobAtticus commented on plugins/outputs/postgresql/postgresql.go:
```diff
@@ -103,149 +130,64 @@ func (p *Postgresql) SampleConfig() string { return sampleConfig }
 	}
 
-	return nil
+	missingColumns, err := p.tables.FindColumnMissmatch(tableName, details)
```
`Missmatch` -> `Mismatch`

RobAtticus commented on plugins/outputs/postgresql/column_details.go:
```diff
@@ -0,0 +1,192 @@
+// and depending on 'tagsAsFK', 'tagsAsJSON', and 'fieldsAsJSON' generate the
+// desired columns (their name, type and which role they play) for both the
+// main metrics table in the DB, and if tagsAsFK == true for the tags table.
+func makeTargetColumns(
```
Something about the complexity of this function makes me think that an interface might be helpful here, where each combo of flags is a different implementation of the interface. But maybe that is overkill, it’s just that the function is a bit tough to digest.

I don’t have anything concrete to offer on that front though.

atanasovskib replied:
I extracted two of the for loops as functions and it greatly reduced the strain of reading the function. Also moved it to an interface, so it can be mocked now. But I didn't know how to simplify the map[bool]map[bool]map[bool]func since we would just be substituting 8 functions for 8 structs

RobAtticus commented on plugins/outputs/postgresql/postgresql.go:
```diff
@@ -2,27 +2,40 @@ package postgresql
-	fieldsJSONColumn = "fields"
+	timeColumnName          = "time"
+	timeColumnDataType      = "timestamptz"
+	timeColumnDefinition    = "time timestamptz"
```
maybe change this to be a concatenation of the two above?
```
timeColumnDefinition = timeColumnName + “ “ + timeColumnDataType
```

atanasovskib commented:
@RobAtticus I would like to merge this also :)

RobAtticus commented on plugins/outputs/postgresql/columns/column_mapper.go:
```diff
@@ -0,0 +1,68 @@
+	}
+}
+
+// Iterate through an array of 'metrics' visiting only those indexed by 'indices'
```
`Target iterates through…`

RobAtticus commented on plugins/outputs/postgresql/utils/utils.go:
```diff
@@ -0,0 +1,168 @@
+	switch value.(type) {
+	case bool:
+		return PgBool
+	case uint64:
```
not strictly needed but since you can’t do fallthrough you can do:
```go
case uint64, int64, int, uint:
  return PgInt8
case uint32, int32:
  return PgInt4
case float64, float32:
  return PgFloat8
```

can ask @iand if that is good style or not

iand replied:
I'd prefer it this way

RobAtticus reviewed (approved):
One doc fix and one potential style, otherwise lgtm

1848 commented on plugins/outputs/postgresql/README.md:
```diff
@@ -1,36 +1,44 @@
-
-  ## Use jsonb datatype for fields. Default is true.
-  # fields_as_jsonb = false
+    address = "host=localhost user=postgres sslmode=verify-full"
```
1) Maybe you want to rename this to "connection" since this is not only the address?
2) I tried to set "sslcompression=1" but it seems like this keyword is unknown. Had no time to debug, should this work?

1848 commented:
Telegraf will fail to write to the output when the database is restarted.
Could you implement some reconnect logic on write error?

Commit:
Attempt reconnect to db if conn is lost and support PG env variables

atanasovskib commented:
Thank you @1848 for testing. Added another commit where the plugin will attempt to reconnect if the connection was lost to the server. Also
* added support for PG environment variables, no longer required to put them in the connection string
* address renamed to connection

About `sslcompression`, since it's not considered safe, it is not supported by the tls implementation in the `crypto/tls` go package. You can check out https://golang.org/src/crypto/tls/common.go (search for `// TLS compression types.`) And the pgx driver we're using relies on this package for establishing the TLS connections

iand commented on plugins/outputs/postgresql/postgresql.go:
```diff
@@ -2,27 +2,32 @@ package postgresql
+	tagCache                    tagsCache
+	rows                        transformer
+	columns                     columns.Mapper
+	dbConnLock                  *sync.Mutex
```
It's conventional to group the mutex above the fields it is intended to protect. It would be useful to add a comment to say which fields must be accessed only when holding the mutex lock.

Also the mutex would be better as a value not a pointer. Using a pointer means copies of the `Postgresql` type will share the same mutex which would be a bad situation.

1848 replied:
@blagojts, thank you very much.

> About `sslcompression`, since it's not considered safe, it is not supported by the tls implementation in the `crypto/tls` go package.

Oh thats unfortunate... I dont care about security in this case. sslcompression is (afaik) the only way to enable compression for the tcp connection. Thanks for the info.

iand commented on plugins/outputs/postgresql/postgresql.go:
```diff
@@ -31,50 +36,72 @@ func init() {
 // Connect establishes a connection to the target database and prepares the cache
 func (p *Postgresql) Connect() error {
-	db, err := newDbWrapper(p.Address)
+	p.dbConnLock = &sync.Mutex{}
```
Making `dbConnLock` a value instead of a pointer means you don't have to do this initialisation. Currently one goroutine could Connect and then call a method that takes the lock, but another goroutine could call Connect and replace the mutex.

iand commented on plugins/outputs/postgresql/postgresql.go:
```diff
@@ -31,50 +36,72 @@ func init() {
 
 // Close closes the connection to the database
 func (p *Postgresql) Close() error {
+	p.tagCache = nil
```
Are any of these fields protected by the mutex?
